examples/amm/src/contract.rs

```diff
@@ -141,24 +136,110 @@ impl Amm {
-                if token0_amount == 0 || token1_amount == 0 {
+                if max_token0_amount == Amount::ZERO || max_token1_amount == Amount::ZERO {
                     return Err(AmmError::NoZeroAmounts);
                 }
 
                 let balance0 = self.get_pool_balance(0).await?;
                 let balance1 = self.get_pool_balance(1).await?;
 
-                if balance1 > 0
-                    && abs_diff_ne!(
-                        balance0 as f64 / balance1 as f64,
-                        token0_amount as f64 / token1_amount as f64,
-                        epsilon = 0.01
-                    )
-                {
-                    return Err(AmmError::BalanceRatioAlteredError);
+                let token0_amount;
+                let token1_amount;
+                if balance0 > Amount::ZERO && balance1 > Amount::ZERO {
+                    let balance0_bigint = BigUint::from_u128(u128::from(balance0))
+                        .expect("Couldn't generate balance0 in bigint");
+                    let balance1_bigint = BigUint::from_u128(u128::from(balance1))
+                        .expect("Couldn't generate balance1 in bigint");
+                    let max_token0_amount_bigint =
+                        BigUint::from_u128(u128::from(max_token0_amount))
+                            .expect("Couldn't generate max_token0_amount in bigint");
+                    let max_token1_amount_bigint =
+                        BigUint::from_u128(u128::from(max_token1_amount))
+                            .expect("Couldn't generate max_token1_amount in bigint");
+
+                    // This is the formula to maintain the ratio:
+                    //      balance0 / balance1 = (balance0 + max_token0_amount) / (balance1 + token1_amount)
+                    //      balance0 * (balance1 + token1_amount) = balance1 * (balance0 + max_token0_amount)
+                    //      balance0 * balance1 + balance0 * token1_amount = balance1 * balance0 + balance1 * max_token0_amount
+                    //      balance0 * token1_amount = balance1 * max_token0_amount
+                    //      token1_amount = (balance1 * max_token0_amount) / balance0
+                    //
+                    // For token0_amount, it would be this:
+                    //      token0_amount = (balance0 * max_token1_amount) / balance1
+
+                    let token0_amount_bigint = (balance0_bigint.clone()
+                        * max_token1_amount_bigint.clone())
+                        / balance1_bigint.clone();
+                    let token1_amount_bigint = (balance1_bigint.clone()
+                        * max_token0_amount_bigint.clone())
+                        / balance0_bigint.clone();
+
+                    if token0_amount_bigint > max_token0_amount_bigint
+                        && token1_amount_bigint > max_token1_amount_bigint
+                    {
+                        return Err(AmmError::BalanceRatioAlteredError);
```

Comment:
You're saying this will never be true? Could you explain why? I can't see it right away

Reply:
Assume both were true. Then:
```
max_token0_amount_bigint
< token0_amount_bigint
= balance0_bigint * max_token1_amount_bigint / balance1_bigint
< balance0_bigint * token1_amount_bigint / balance1_bigint
= balance0_bigint * balance1_bigint * max_token0_amount_bigint / balance0_bigint / balance1_bigint
= max_token0_amount_bigint
```

So `max_token0_amount_bigint < max_token0_amount_bigint`, a contradiction.

Reply:
More intuitively:
Either `max_token0_amount / balance0` is greater than `max_token1_amount / balance1`, or less, or equal.
If it's greater, you are offering more token0 than necessary, so we compute the actual token 0 amount from your `max_token1_amount`. If it's less, it's the other way round. If they're exactly equal, we can use both `max_token0_amount`  and `max_token1_amount` directly, because they already are in exactly the right ratio.

Reply:
I see, got it! Thank you